### PR TITLE
Add Puck

### PR DESCRIPTION
**Pull request by EMattfolk**
Haha, gaming.

### CODE REVIEW

File: game.sy [hunk 1]
```diff
@@ -99,8 +99,75 @@ player_new :: fn pos: (float, float) -> Player {
 
 :Player is :Entity
 
+Puck :: blob {
+    pos: (float, float),
+    rot: float,
+    vel: float,
+    acc: float,
+    holder: Player | void,
```
Comment (sornas) [inline game.sy]:
```suggestion
    holder: Player?,
```

File: game.sy [hunk 2]
```diff
@@ -99,8 +99,75 @@ player_new :: fn pos: (float, float) -> Player {
...
+    hold_cooldown: float,
+    draw: fn -> void,
+    process_input: fn Input, float -> void,
+    update: fn float -> void,
+}
+
+puck_new :: fn pos: (float, float) -> Puck {
+    self: !Puck = nil
+    self = Puck{
```
Comment (sornas) [inline game.sy:146]: Not really related to this PR but is this something we want to be able to do without the nil? @FredTheDino

File: game.sy [hunk 3]
```diff
@@ -99,8 +99,75 @@ player_new :: fn pos: (float, float) -> Player {
...
+        holder: nil,
+        hold_cooldown: 0.0,
+        draw: fn -> void {
+            rgb :: (1.0, 1.0, 1.0)
+            l_gfx_rect' self.pos, (0.04, 0.04), self.rot, rgb
+            l_gfx_rect' self.pos, (0.04, 0.04), self.rot + 3.14 / 3.0, rgb
+            l_gfx_rect' self.pos, (0.04, 0.04), self.rot + 3.14 * 2.0 / 3.0, rgb
+        },
+        process_input: fn input: Input, delta: float -> void {
+            if input.shoot && self.holder is :Player {
```
Comment (sornas) [inline game.sy]: Now that's expressive!

File: game.sy [hunk 4]
```diff
@@ -99,8 +99,75 @@ player_new :: fn pos: (float, float) -> Player {
...
+                holder: !Player = self.holder
```
Comment (sornas) [inline game.sy]: This, not so much :(

**Review by sornas — APPROVED** [review]

**Comment by sornas** [comment]
Gaming!

File: game.sy [hunk 5]
```diff
@@ -99,8 +99,75 @@ player_new :: fn pos: (float, float) -> Player {
 
 :Player is :Entity
 
+Puck :: blob {
+    pos: (float, float),
+    rot: float,
+    vel: float,
+    acc: float,
```
Comment (FredTheDino) [inline game.sy:136]: Is this how we want to represent movement? Feels like we want a lot of forces acting on this - and tuple addition is defined in the language.

File: game.sy (hunk 2 above)
Comment (FredTheDino) [reply game.sy:146]: Yes - I want to add a special case in the compiler for it.

File: game.sy [hunk 6]
```diff
@@ -124,9 +192,13 @@ init :: fn {
     push' players, player
     push' entities, player
 
+    puck.holder = player
+
     player := player_new' (1.0, 0.0)
     push' players, player
     push' entities, player
+
+    push' entities, puck
```
Comment (FredTheDino) [inline game.sy:218]: nice

File: game.sy [hunk 7]
```diff
@@ -99,8 +99,75 @@ player_new :: fn pos: (float, float) -> Player {
...
+            l_gfx_rect' self.pos, (0.04, 0.04), self.rot + 3.14 / 3.0, rgb
+            l_gfx_rect' self.pos, (0.04, 0.04), self.rot + 3.14 * 2.0 / 3.0, rgb
+        },
+        process_input: fn input: Input, delta: float -> void {
+            if input.shoot && self.holder is :Player {
+                holder: !Player = self.holder
+                self.vel = holder.vel * 2.0
+                self.rot = holder.rot
+                self.holder = nil
+                self.hold_cooldown = 0.1
```
Comment (FredTheDino) [inline game.sy]: Shouldn't this logic be placed on the player? Do we really want to give input to the puck?

**Review by FredTheDino — APPROVED** [review]

File: game.sy (hunk 7 above)
Comment (EMattfolk) [reply game.sy]: Yes, I am not sure what I was thinking

**Pull request merged**